src/syntaxtree.c

```diff
@@ -158,6 +185,74 @@ static Obj SyntaxTreeDefaultCompiler(Obj result, Expr expr)
     return result;
 }
 
+// TODO: use the following as start of a "reverse compiler"
+static Expr SyntaxTreeDefaultCoder(Obj node)
+{
+
+    Obj  typeString = ElmPRec(node, RNamName("type"));
+    UInt tnum = FIND_TNUM_FROM_STRING(CSTR_STRING(typeString));
+
+    if (tnum == LAST_EXPR_TNUM + 2) {
+        ErrorQuit("Unrecognized syntax tree node type", 0, 0);
+    }
+
+    CompilerT comp = Compilers[tnum];
+
+    if (comp.code != SyntaxTreeDefaultCoder) {
+        return comp.code(node);
+    }
+
+    // compute how many slots we need
+    UInt slots = comp.arity;
+    UInt arity = comp.arity;
+
+    UInt isvararg = comp.arity > 0 && comp.args[comp.arity - 1].argcomp == 0;
+    Obj  vararglist;
+
+    if (isvararg) {
+        arity--;
+        vararglist = ElmPRec(node, RNamName(comp.args[arity].argname));
```

Comment:
Let's use `ELM_REC` everywhere instead of `ElmPRec` to get free input validation.

Reply:
Done